### ISSUE TITLE
Efficient Frontier does not work for Bounds 0 to 1 and a List of 6000 Assets

### ISSUE THREAD

**Issue by AdamTheDestroyer** [robertmartin8/PyPortfolioOpt#128]
Efficient Frontier as shown in your tutorial does not work for Bounds 0 to 1 and a List of 6000 Assets.

I even cant show you a Error Message, because i did not get one. 
When i switch to Bounds -1 and 1 it works.
I tryed it with less Assests and Bounds 0 to 1 and that worked well.

Thank you!

**Comment by robertmartin8** [comment]
Hi @AdamTheDestroyer 

Thanks for raising the issue! When you say it doesn't work, do you mean that the interpreter crashed? or did it just not do anything?

To be honest, I've never tested PyPortfolioOpt with more than a few hundred assets, so I don't know how well it performs on large portfolios. I can imagine that the optimiser might have more trouble, but it's weird that it fails silently. 

If you don't mind me asking, what are you trying to achieve out of optimising a portfolio of 6000 assets? My view is that PyPortfolioOpt should not be used for asset selection, just for the optimisation part. Which is why it has mostly been tested for portfolios of 10-100 stocks.

Best,
Robert

**Comment by AdamTheDestroyer** [comment]
Hi Robert,

Yes I know I do some strange things :-)
First Step is to optimize my Stocks Daily and the 2nd Step is to do some checks like, : how much is the new weight higher/lower than yesterday (for example in dollar 3000$), does it cross some Chart technic borders and soo on...

On these things I wrote some backtesting code (just to compare some portfolios against each other) in some months.
Thats the reason why I want to optimize and compare such a big portfolio.

I tried it again and now I have a Error message (sorry for my misleading post before):
SolverError                               Traceback (most recent call last)
<ipython-input-1-2c1bb21f7bd7> in <module>
     45         # Optimise for maximal Sharpe ratio
     46         ef = EfficientFrontier(mu, S, weight_bounds=(0,1))
---> 47         raw_weights = ef.max_sharpe()
     48 
     49 

~\anaconda3\lib\site-packages\pypfopt\efficient_frontier.py in max_sharpe(self, risk_free_rate)
    211         ] + new_constraints
    212 
--> 213         self._solve_cvxpy_opt_problem()
    214         # Inverse-transform
    215         self.weights = (self._w.value / k.value).round(16) + 0.0

~\anaconda3\lib\site-packages\pypfopt\base_optimizer.py in _solve_cvxpy_opt_problem(self)
    194         try:
    195             opt = cp.Problem(cp.Minimize(self._objective), self._constraints)
--> 196             opt.solve()
    197         except (TypeError, cp.DCPError):
    198             raise exceptions.OptimizationError

~\anaconda3\lib\site-packages\cvxpy\problems\problem.py in solve(self, *args, **kwargs)
    288         else:
    289             solve_func = Problem._solve
--> 290         return solve_func(self, *args, **kwargs)
    291 
    292     @classmethod

~\anaconda3\lib\site-packages\cvxpy\problems\problem.py in _solve(self, solver, warm_start, verbose, parallel, gp, qcp, **kwargs)
    573         full_chain = self._solving_chain.prepend(self._intermediate_chain)
    574         inverse_data = self._intermediate_inverse_data + solving_inverse_data
--> 575         self.unpack_results(solution, full_chain, inverse_data)
    576         return self.value
    577 

~\anaconda3\lib\site-packages\cvxpy\problems\problem.py in unpack_results(self, solution, chain, inverse_data)
    716             raise error.SolverError(
    717                     "Solver '%s' failed. " % chain.solver.name() +
--> 718                     "Try another solver, or solve with verbose=True for more "
    719                     "information.")
    720         self.unpack(solution)

SolverError: Solver 'OSQP' failed. Try another solver, or solve with verbose=True for more information.

**Comment by robertmartin8** [comment]
Hi Adam,

Sorry for the slow response. I'm don't currently have any solutions for you – if you provide a well-posed problem but it is just too big for the solver, the only real solution is to use a different solver. In the next release, I will make it possible to modify the solver so that users can have more details about convergence etc.

On a practical note, to optimise for such large portfolios, it is often advisable to use a factor model (though I haven't implemented this yet).

In the meantime, I will leave this issue up in case anyone else has some insight. 

Robert

**Comment by robertmartin8** [comment]
As of v1.2.3, it's possible to change the solvers by assigning `ef.solver = "..."` before optimising. It might be the case that one of the other solvers (see a list [here](https://www.cvxpy.org/tutorial/advanced/index.html#choosing-a-solver)) can handle the large problem

**Comment by AdamTheDestroyer** [comment]
Hi @phschiele and @robertmartin8,
i wish you a happy and healthy new year 2021!
@phschiele i have to ask my client i am working with, because of possible business secretes. After that i maybe can load up a testfile. But until that i can descibe my Data: I am using only closing prices and Datasets from 2016 until today. Stocks which arent that old, are filled up from 2016 with their individual first stock price. First column is the Date column, and after that column there are the closing prices of over 5000 Stocks ("Date","1&1DrillischAG","1-800-FlowerscomInc","11880SolutionsAG","1300SmilesLtd","180DegreeCapitalCorp",.....n5000)

I hope i descibed my Data well.

Greetings,
Adam

**Comment by robertmartin8** [comment]
@AdamTheDestroyer portfolio optimisation is difficult when you have more assets than datapoints. 

In your case, you have ~250 * 4 = 1000 datapoints, but 5000 stocks. This is likely to lead to an ill-conditioned covariance matrix, so I am not surprised that optimisation is failing.

Minor housekeeping note, please keep followups in this thread. I will reopen this, but definitively closing the more recent thread in #239.

**Comment by phschiele** [comment]
@AdamTheDestroyer You might find [this post](https://stats.stackexchange.com/a/60631) relevant to see why it might be hard to perform the optimization with `number of assets >> number of observations`. 
Would nevertheless be interesting to take a look at the data. If it would make things easier for you, it would be fully sufficient for me if you replace the names/ISINs of the stocks by `1,2,...`.

**Comment by AdamTheDestroyer** [comment]
[Stocks.zip](https://github.com/robertmartin8/PyPortfolioOpt/files/5783488/Stocks.zip)

@phschiele So here we go. 
Thats an large testfile with over 6000 Stocks, with continued numbers as columnnames

@robertmartin8 I will try to reduce my number of stocks as you mentioned (ignoring negative returns) and hope that that is the key. Till now my first step was portfolio optimization and in a second step i did the stock selection with standard things like looking at some moving average signs...

**Comment by phschiele** [comment]
Hi @AdamTheDestroyer,

I hope the selection tips by @robertmartin8  already helped you to solve your problem. Indeed, I think optimizing on all stocks and later doing the selection does not seem intuitive to me.

Nevertheless, I took a look at your csv file. I am currently writing a solver for portfolio optimization problems and wanted to check how it holds up for a large number of assets. I selected the first `n` assets and let `n` got to the full sample size, calculating the full efficient frontier for each (not just a single point on the frontier).
Indeed, it seems to perform well:
![image](https://user-images.githubusercontent.com/44360364/105543407-0aefd000-5cfa-11eb-8d71-ff8d9a59ccf7.png)

The resulting efficient frontier also looks reasonable (again, speaking from a technical perspective).
![image](https://user-images.githubusercontent.com/44360364/105543499-22c75400-5cfa-11eb-9a6c-4c3640b511a3.png)

Sadly, I had to postpone the release of the solver due to other projects, but I hope to eventually get back to it and release here ;)
Thanks for lending me the real world sample!

Also relevant for #263.

Cheers.

**Comment by AdamTheDestroyer** [comment]
Wow, that sounds impressive to me. Is there any possibility to help you in this project? I would be very glad if we could reach this massive improvement. I spent more than two month to code a workaround with a Monte Carlo solution. Even Monte Carlo took 4-5 hours for 100k times x 6000 assets random weights + Efficient Frontier. Therefore, I took a deep dive into parallel computing and vectorizing things, so that in the end 90% is done with NumPy instead of Pandas (It just seems easy). I even programmed the exponential moving covariance for parallel computing to improve the performance (because the Pandas solution from this PyPortfolioOpt project was much too slow for 6000 assets).

Long story short: I would be very glad if you have  time for publishing your great improvement.

Greetings Adam

**Comment by kayuksel** [comment]
This is the winner of Large-Scale Global Optimization Competition in 2018 and it is claimed to be the fastest one among them:
https://github.com/dmolina/shadeils

Enclosed are results from my own large universe optimization that is done with my proprietary optimizer, which I plan to publish.
[backtest.pdf](https://github.com/robertmartin8/PyPortfolioOpt/files/5976051/backtest.pdf)

**Comment by phschiele** [comment]
@AdamTheDestroyer Many thanks for your kind offer. I will reach out if there is anything I need help with. Glad to hear you found a way to optimize the portfolio through MC!

@kayuksel Thanks for the reference. At first glance, it looks like this is a general purpose optimizer, Would be interesting to see it applied to portfolio optimization. 
Also looking forward to read about your optimizer/strategy!

**Comment by robertmartin8** [comment]
@kayuksel Looking forward to the publication of your optimiser! I think it's truly a good thing for the ecosystem to have different tools that are suited to different scenarios. PyPortfolioOpt was primarily designed for small-ish portfolios (<100), so it would be great to have more tooling that can be applied to large universes.
